bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCallbackUtils.java

```diff
@@ -0,0 +1,311 @@
+         * @param rc
+         *            return code
+         * @param ctx
+         *            optional control object
+         */
+        @Override
+        public void deleteComplete(int rc, Object ctx) {
+            finish(rc, null, future);
+        }
+    }
+
+    static class LastAddConfirmedCallback implements AsyncCallback.AddLacCallback {
+        static final LastAddConfirmedCallback INSTANCE = new LastAddConfirmedCallback();
+        /**
+         * Implementation of callback interface for synchronous read method.
+         *
+         * @param rc
+         *          return code
+         * @param lh
+         *          ledger identifier
+         * @param ctx
+         *          control object
+         */
+        @Override
+        public void addLacComplete(int rc, LedgerHandle lh, Object ctx) {
+            if (rc != BKException.Code.OK) {
+                LOG.warn("LastAddConfirmedUpdate failed: {} ", BKException.getMessage(rc));
+            } else {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Callback LAC Updated for: {} ", lh.getId());
+                }
+            }
+        }
+    }
+
+    static class SyncReadCallback implements AsyncCallback.ReadCallback {
+
+        private final CompletableFuture<Enumeration<LedgerEntry>> future;
+
+        public SyncReadCallback(CompletableFuture<Enumeration<LedgerEntry>> future) {
+            this.future = future;
+        }
+
+        /**
+         * Implementation of callback interface for synchronous read method.
+         *
+         * @param rc
+         *          return code
+         * @param lh
+         *          ledger handle
+         * @param seq
+         *          sequence of entries
+         * @param ctx
+         *          control object
+         */
+        @Override
+        public void readComplete(int rc, LedgerHandle lh,
+                                 Enumeration<LedgerEntry> seq, Object ctx) {
+            finish(rc, seq, future);
+        }
+    }
+
+    static class FutureReadResult
+        extends CompletableFuture<Iterable<org.apache.bookkeeper.client.api.LedgerEntry>>
+        implements AsyncCallback.ReadCallback {
+
+        /**
+         * Implementation of callback interface for read method of {@link ReadHandle}.
+         *
+         * @param rc
+         *          return code
+         * @param lh
+         *          ledger handle
+         * @param seq
+         *          sequence of entries
+         * @param ctx
+         *          control object
+         */
+        @Override
+        @SuppressWarnings("unchecked")
+        public void readComplete(int rc, LedgerHandle lh,
+                                 Enumeration<LedgerEntry> seq, Object ctx) {
+            if (rc != BKException.Code.OK) {
+                this.completeExceptionally(BKException.create(rc).fillInStackTrace());
+            } else {
+                this.complete((Iterable) () -> Iterators.forEnumeration(seq));
```

Comment:
I am a bit confused about this change.

you changed from `Iterators.forEnumeration(..)` to `() -> Iterators.forEnumeration(..)`. why?

Reply:
Because we need an Iterable and not an Iterator. Tests started to fail actually